### ISSUE TITLE
Using multiple datasets for fastMNN

### ISSUE THREAD

**Issue by piyushjo15** [LTLA/batchelor#15]
Hi,

I am trying to use multiple datasets for input as fastMNN. I can run it successfully with the below code.

`mnn.out <- fastMNN(sce1,sce2)`

As I have more than 10 datasets, and it will be an hassle to de type their names. Secondly, the batch variable is a numeric array rather than character array in the output.

I tried following the OSCA tutorial way to first create a list, process the individual elements of list and then use that as input. 
`list.sce <-list(A=sce1,B=sce2)`
`mnn.out <- fastMNN(list.sce)`

But, this gave me an error that I need to provide batch information. However, in the example in tutorial
`mnn.pancreas <- fastMNN(normed.pancreas)`
There is no batch information given and still the batch appear as characters. 

When I provided batch information as character or factor array, both time it gave me an error regarding difference in dimensions. I created the array repeating sample ID times the number of cells in each sample.

Could you suggest what could be the solution?

**Comment by LTLA** [comment]
Off the top of my head, there are three ways to pass multiple batches into **batchelor** functions:

1. `FUN(batch1, batch2, batch3, batch4, ...)` and so on, where each `batchX` is a matrix or `SingleCellExperiment` object containing all cells from a single batch.
2. `FUN(all.batches, batch=batch.id)` where `all.batches` is a matrix or SCE containing all cells from all batches, and `batch.id` is a vector specifying which cell belongs to which batch.
3. `FUN(list(batch1, batch2, batch3, batch4, ...))`, which is basically the same as **1** but you can make your own list. Currently only available in BioC-devel (1.3.14).

I suspect you're probably using the current BioC-release version, for which **3** isn't yet available. But that's okay, because you can just use **1** with a neat trick:

```r
do.call(FUN, list(batch1, batch2, batch3, batch4, ...))
```

Slightly more typing, which is why I added **3**, but it does the same thing.

**Comment by piyushjo15** [comment]
Thanks Aaron. Also when combining SCE objects using cbind() it gives an error regarding some entries not being similar. One thing that was done in OSCA tutorial was to assign rowData of one object to the other one. So do I have to set all rowData to same (may be NULL?).

**Comment by piyushjo15** [comment]
Thanks Aaron. Now I used the `rlist:: list.append` that also helps me pass the arguments. I also tried combining the datasets using noCorrect(), into a one SCE. Does it only combine the logcounts from all the datasets? This is what is shown in the OSCA tutorial.

**Comment by LTLA** [comment]
That's correct, `noCorrect()` will only _correct_ the assays being used in the correction. This mirrors the behavior of the other correction methods, which operate on just one assay type in any one call.

If you want to _combine_ (i.e., `cbind`) but not _correct_ the other assays, you might consider using `correctExperiments()`. This will correct one of the assays but combines the the other assays, making it useful to have both the original and corrected assays combined in the same object.

**Comment by piyushjo15** [comment]
So `noCorrect` will perform corrections? Or by "correct" you mean just merge _logcounts_?

I am running the `correctExperiments()` as below, I guess this is merging and also performing FastMNN based correction? The output is also giving a _reconstructed_ counts.
```
com.sce <- do.call(
  correctExperiments,
  list.append(norm.sce,subset.row = com.hvg,
              PARAM=FastMnnParam(d = 50, k = 20,
                                 BSPARAM=BiocSingular::RandomParam(deferred=TRUE))))
```
I am getting this warning, which I wasn't getting in `noCorrect`. I saw it is related with genes but don't know what it is.
```
Warning message:
In (function (..., batch = NULL, restrict = NULL, subset.row = NULL,  :
  ignoring non-identical 'rowRanges' fields
```


Second point: you mentioned that you use unscaled normalized counts from individual datasets for marker analysis. That's what I was doing with `noCorrect()`. So I made another SCE from `multiBatchNorm()`, _norm.sce_, which I used for 'fastMNN()'. Then I just added the corrected MNN dimensions for each cell to the merged object from `noCorrect()`.
For correctedExperiments, do I provid the output of `multiBatchNorm()`, or it is performing `multiBatchNorm()` during the run and that is what the _reconstructed_ assay is, the rescaled _logcounts_ for each cell? 

Last point: How come merging different SCE object is not throwing error for row names, which are barcode ids from individual datasets. Barcode ids from different runs would still be similar, and in Seurat it modifies it with prefix, but here it not. I checked there were duplicated names present in `row.names(colData(com.sce))`.

Thanks a lot for your explanations!!

**Comment by LTLA** [comment]
> So `noCorrect` will perform corrections? Or by "correct" you mean just merge _logcounts_?

Yes, that's right. `noCorrect`'s "correction" is to not do any correction! What I should have said is that all of the `*Correct` functions (and `fastMNN`) will only operate on the specified assay. In `noCorrect`'s case, the correction algorithm is... to do nothing, in which case, it is the same as merging them via a standard `cbind`.

> I am running the `correctExperiments()` as below, I guess this is merging and also performing FastMNN based correction?

Yep.

> I am getting this warning, which I wasn't getting in `noCorrect`. I saw it is related with genes but don't know what it is.

Your individual objects have some differences in the row-level metadata. Possibly they have different gene coordinates (e.g., if they were aligned to different genome builds) or they have differences in various metadata fields (e.g., changes in gene symbol across Ensembl versions). This may or may not be concerning; if you're using Ensembl or Entrez IDs as the row names, there is no need to worry as each ID refers unambiguously to a constant thing. If you're using gene symbols... well, you'll just have to hope that the different objects are referring to the same gene when they are using the same symbol.

> For correctedExperiments, do I provid the output of `multiBatchNorm()`, or it is performing `multiBatchNorm()` during the run and that is what the _reconstructed_ assay is, the rescaled _logcounts_ for each cell?

Right now, `correctExperiments` doesn't run `multiBatchNorm()`. I suppose I could add an option to do it automatically, which might make life a little easier for you.

Regardless of whether that happens, the `reconstructed` assay is _not_ the rescaled log-counts. I compute them by taking the corrected PC scores and projecting them back into the original expression space. So you can (sort of) think of them as the expression values after correction; though I wouldn't use them for anything other than visualization, see comments [here](https://osca.bioconductor.org/integrating-datasets.html#using-corrected-values).

> Last point: How come merging different SCE object is not throwing error for row names, which are barcode ids from individual datasets. Barcode ids from different runs would still be similar, and in Seurat it modifies it with prefix, but here it not. I checked there were duplicated names present in `row.names(colData(com.sce))`.

You can have duplicate column and row names in many R objects. For example:

```r
X <- matrix(0, 10, 10)
colnames(X) <- rep("A", 10)
rownames(X) <- rep("B", 10)
```

If necessary, disambiguation is possible via the `batch` field in the column metadata. I guess I _could_ modify the column names, but I don't do so to be consistent with how users expect things to behave. Specifically, R doesn't bother to modify duplicate names for matrices:

```r
colnames(cbind(X, X))
##  [1] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
## [20] "A"
```

Not even for `data.frame`s:

```r
Y <- data.frame(A=1:10)
cbind(Y, Y)
##     A  A
## 1   1  1
## 2   2  2
## 3   3  3
## 4   4  4
## 5   5  5
## 6   6  6
## 7   7  7
## 8   8  8
## 9   9  9
## 10 10 10
```

**Comment by piyushjo15** [comment]
Hi Aaron,

I don't think you have to add `multiBatchNorm()` in the `correctExperiments()` run. I just wanted to clarify that I have to provide the rescaled batch-normalized list of SCEs as input to `correctExperiments()`.

I have aligned the sequences to the same genome build, so all my genes should have consistent symbols and ids. I am using gene symbols (rather than ids). That's why I was wondering why am I getting the warning, if  didn't get any similar warning in `noCorrect()`. May be one of the assay or meta.data object being merged in `correctExperiments()` is getting affected. Shall I just ignore the warning, hoping it is not affecting any important data matrix (counts, logcounts) in the combined SCE object.

**Comment by LTLA** [comment]
Probably some of the batch-specific metadata, I would guess. If you stored, say, the average expression or the proportion of expressing cells in the `rowData()`, those are unlikely to be the same across batches, thus triggering the warning. 

In any case, I have just modified `correctExperiments()` to be (i) more relaxed about keeping row metadata fields if they do not conflict, and (ii) to report those conflicting fields to make it easier to spot problems. This should be available in **batchelor** 1.1.16.
